### PR TITLE
Clean up memories when syncing

### LINKED ISSUE #1576: High memory usage while syncing
During the block synchronization, codechain uses main memory over 8GB.
And the occupied memory was not cleared after the synchronization finished.

### PR DESCRIPTION
**Pull request by HoOngEe**
The commits in this pr piled upon the #1567
It closes #1576

### CODE REVIEW

**Comment by sgkim126** [comment]
@HoOngEe Would you shrink `targets` too?

**Comment by HoOngEe** [comment]
Before shrinking `targets` the memory usage reduced from about `9.0GB` to `5.3GB`.

**Comment by HoOngEe** [comment]
I ran memory profiling tool valgrin --tool=massif, used massif-visualizer and got the following graph.
memory_profile

This time I got data of only 68000 blocks synchronization. So I cannot certain that the data from these two key parts: `header_queue` of importer, `downloaded` of `HeaderDownloader` would be shrunk later. The running time increased about 20 times before attaching heap memory profiler so that I cannot observe in full-time domain. Now I'll check `header_queue` and `body_queue`

**Comment by HoOngEe** [comment]
`processing: RwLock<HashMap<H256, U256>>` also should be shrinked

File: sync/src/block/downloader/body.rs [hunk 1]
```diff
@@ -111,6 +115,7 @@ impl BodyDownloader {
                 break
             }
         }
+        self.downloaded.shrink_to_fit();
         self.targets.drain(0..result.len());
```
Comment (sgkim126) [inline sync/src/block/downloader/body.rs]: Please shrink `targets` too.

Comment (HoOngEe) [reply sync/src/block/downloader/body.rs]: locally I shrank `targets`. In next force-push the change will be included.

**Comment by HoOngEe** [comment]
I got a full-time synchronization data through last weekends.
full-time

The purple area, which was allocated in the `block_header_data` of `HeaderChain` implementing `HeaderProvider`, accounts for almost memory uncleared even after synchronizing old blocks.
It uses `HashMap` named `header_cache` internally. But there is no entry clearing routine for this.

**Comment by sgkim126** [comment]
@HoOngEe I know still there is a memory to cleaning up. However, I think it's hard to fix the problem, there is no apparent time to clean the cache. It should change the implementation to LRU like implementation.
So I propose to merge this PR and keep #1576 open until we fix the memory issue completely.
How about you?

**Comment by HoOngEe** [comment]
@sgkim126 Ok, let's merge this pr now and fix the cache implementation later.

File: core/src/verification/queue/mod.rs [hunk 2]
```diff
@@ -429,6 +430,7 @@ impl<K: Kind> VerificationQueue<K> {
                 *td -= score;
             }
         }
+        processing.shrink_to_fit();
```
Comment (sgkim126) [inline core/src/verification/queue/mod.rs]: This seems redundant.

Comment (HoOngEe) [reply core/src/verification/queue/mod.rs]: Yes, It's redundant. I fixed it. Thank you.

**Comment by sgkim126** [comment]
@HoOngEe LGTM. Please apply it to rc-1.4.0 too.

**Pull request merged**